File: Code/petiapp/lib/models/post.dart
import 'location.dart';
import 'pet.dart';

class Post {
  Post({
    this.file,
    this.content,
    this.createdAt,
    this.id,
    this.likes,
    this.comments,
    this.creator,
    this.isLiked,
    this.location,
  });

  FileObject? file;
  dynamic content;
  String? createdAt;
  String? id;
  int? likes;
  bool? isLiked;
  int? comments;
  Pet? creator;
  Location? location;

  bool isDeleted = false;

  factory Post.fromMap(Map<String, dynamic> json) => Post(
        file: FileObject.fromMap(json["file"]),
        content: json["content"],
        isLiked: json["is_liked"],
        createdAt: json["created_at"],
        id: json["id"],
        likes: json["likes"],
        comments: json["comments"],
        location: json["location"] == null
            ? null
            : Location.fromMap(json["location"]),
        creator: json["creator"] == null ? null : Pet.fromMap(json["creator"]),
      );
}

class FileObject {
  FileObject({
    required this.url,
    this.bgcolor,
    this.thumbinal,
    required this.mimetype,
  });

  String url;
  String? bgcolor;
  String? thumbinal;
  String mimetype;

  factory FileObject.fromMap(Map<String, dynamic> json) => FileObject(
        url: json["url"],
        bgcolor: json['bgcolor'],
        thumbinal: json["thumbinal"],
        mimetype: json["mimetype"],
      );

  bool get isVideo => mimetype.startsWith('video');
}

File: Code/server/app/models/Post.py
from datetime import datetime
from enum import Enum
from typing import Optional
from .engine import engine
from . import Pet, User, Location
from odmantic import Model, Reference, Field

class FileType(str, Enum):
    VIDEO = "video"
    IMAGE = "image"

class Post(Model):
    file: Optional[dict]
    content: Optional[str]
    file_type: Optional[FileType]
    creator: Pet = Reference()
    location: Location = Reference()
    likes: int = 0
    comments: int = 0
    created_at: datetime = Field(default_factory=datetime.utcnow)

    class Config:
        collection = 'posts'
        parse_doc_with_default_factories = True

class Like(Model):
    liked: Post = Reference()
    liker: User = Reference()
    created_at: datetime = Field(default_factory=datetime.utcnow)

    class Config:
        collection = 'likes'
        parse_doc_with_default_factories = True

class Comment(Model):
    content: str
    commented: Post = Reference()
    commenter: User = Reference()
    created_at: datetime = Field(default_factory=datetime.utcnow)

    class Config:
        collection = 'comments'
        parse_doc_with_default_factories = True

class PostFunc():
    async def get_classic_data(post: Post, id):
        is_exist = await engine.find_one(Like, Like.liker == id, Like.liked == post.id)
        if is_exist:
            is_liked = True
        else:
            is_liked = False
        data = post.dict(exclude={'id', 'creator', 'location'})
        data["id"] = str(post.id)
        data["creator"] = post.creator.dict(include={'avatar', 'name', 'gender'})
        data["creator"]["id"] = str(post.creator.id)
        data["creator"]["owner_id"] = str(post.creator.owner.id)
        data["location"] = post.location.dict(exclude={'id', 'owner'})
        data["location"]["id"] = str(post.location.id)
        data["is_liked"] = is_liked
        return data

class CommentFunc():
    async def get_classic_data(comment: Comment):
        data = comment.dict(include={'commenter': {
                'avatar',
                'first_name',
                'last_name',
            }})
        data['content'] = comment.content
        data['created_at'] = comment.created_at
        return data